File: examples/group-video-call/src/utils/turnCredentials.ts
import { IceServer } from '../types';

// Configure these values with your Metered credentials
const METERED_DOMAIN = '';
const METERED_API_KEY = '';

export async function fetchTurnCredentials(): Promise<IceServer[]> {
  const response = await fetch(
    `https://${METERED_DOMAIN}/api/v1/turn/credentials?apiKey=${METERED_API_KEY}`
  );

  if (!response.ok) {
    throw new Error('Failed to fetch TURN credentials');
  }

  const iceServers = await response.json();
  return iceServers;
}

File: examples/flutter-group-video-call/lib/services/turn_credentials.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../models/ice_server.dart';

// Configure these values with your Metered credentials.
// Get them at https://www.metered.ca/stun-turn/
const String meteredDomain = '';
const String meteredApiKey = '';

const IceServer meteredStunFallback =
    IceServer(urls: 'stun:stun.metered.ca:80');

Future<List<IceServer>> fetchIceServers() async {
  if (meteredDomain.isEmpty || meteredApiKey.isEmpty) {
    throw StateError('METERED_DOMAIN / METERED_API_KEY are not configured');
  }
  final uri = Uri.parse(
    'https://$meteredDomain/api/v1/turn/credentials?apiKey=$meteredApiKey',
  );
  final res = await http.get(uri);
  if (res.statusCode != 200) {
    throw HttpException(
      'TURN credential fetch failed (${res.statusCode}): ${res.body}',
    );
  }
  final body = jsonDecode(res.body) as List<dynamic>;
  return body
      .map((e) => IceServer.fromJson((e as Map).cast<String, dynamic>()))
      .toList();
}

class HttpException implements Exception {
  final String message;
  HttpException(this.message);
  @override
  String toString() => message;
}
